### PR TITLE
Initial Implementation of RSC

### PR DESCRIPTION
**Pull request by KhaledEmaraDev**
### Summary

This is an initial implementation of React Server Components (RSC). In this PR:
1. I add the necessary code to render to the RSC wire format using Node.js Streams.
2. I add a hook `useRSC()` and a React Component that extends `react-router-dom`'s `Route` named `RSCRoute` to be used to route to RSC components.
3. I make sure to separate Server and Client functions because sometimes Server code requires somethings only available to Node.

## Related to

https://github.com/shakacode/react_on_rails_pro/pull/346

### Pull Request checklist
_Remove this line after checking all the items here. If the item is not applicable to the PR, both check it out and wrap it by `~`._

- [ ] Add/update test to cover these changes
- [ ] Update documentation
- [ ] Update CHANGELOG file  
  _Add the CHANGELOG entry at the top of the file._

### Other Information

_Remove this paragraph and mention any other important and relevant information such as benchmarks._

- - -
This change is [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1573)

### CODE REVIEW

File: node_package/src/renderReactServerComponent.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  try {
+    const componentObj = ComponentRegistry.get(name);
+    const { component } = componentObj;
+    const reactRenderingResult = React.createElement(component as ReactComponent, props);
+
+    const processReactElement = () => {
+      try {
+        const path = require('path');
+        const { readFileSync } = require('fs');
+        const reactClientManifest = readFileSync(path.resolve(__dirname, 'react-client-manifest.json'), 'utf8');
```
Comment (KhaledEmaraDev) [inline node_package/src/renderReactServerComponent.ts]: This should only be read once and cached.
Shared across workers is better.

**Comment by ahangarha** [comment]
Is this PR ready for review? If yes, May you please update the description to reflect the state of the PR?

**Comment by justin808** [comment]
@KhaledEmaraDev close this one?

**Comment by KhaledEmaraDev** [comment]
@justin808 No, it's needed for the Pro version to work. I think it could be accepted as-is. I will not introduce new changes to it.

File: package.json [hunk 2]
```diff
@@ -92,5 +98,10 @@
   "bugs": {
     "url": "https://github.com/shakacode/react_on_rails/issues"
   },
-  "homepage": "https://github.com/shakacode/react_on_rails#readme"
+  "homepage": "https://github.com/shakacode/react_on_rails#readme",
+  "husky": {
+    "hooks": {
+      "pre-commit": "yalc check"
```
Comment (justin808) [inline package.json:104]: why is this needed pre-commit?

File: package.json [hunk 3]
```diff
@@ -32,6 +32,7 @@
     "eslint-plugin-jsx-a11y": "^6.7.1",
     "eslint-plugin-prettier": "^3.4.1",
     "eslint-plugin-react": "^7.32.1",
+    "husky": "^4.3.6",
```
Comment (justin808) [inline package.json:35]: husky and yalc should be separate PRs.

File: package.json [hunk 4]
```diff
@@ -54,6 +56,10 @@
     "react": ">= 16",
     "react-dom": ">= 16"
   },
+  "exports": {
+    ".": "./node_package/lib/ReactOnRails.js",
+    "./*": "./node_package/lib/*.js"
+  },
```
Comment (justin808) [inline package.json:62]: @KhaledEmaraDev can you explain this?

Seems odd to start exporting everything from package.json

Why not add what's necessary to the current exports?

We already have:
https://github.com/shakacode/react_on_rails/blob/master/node_package/src/ReactOnRails.ts#L292-L293

File: yarn.lock
Comment (justin808) [inline yarn.lock:1]: this file should not be changing, right?

**Review by justin808 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**